peps/pep-0803.rst

```diff
@@ -0,0 +1,321 @@
+builds.
+To allow this, the ``PyObject`` internal structure and related APIs
+will be removed from the Limited API, requiring migration to new API
+for common tasks like defining modules and most classes.
+
+
+Terminology
+===========
+
+This PEP uses “GIL-enabled build” as antonym to “free-threaded build”,
+that is, an interpreter or extension built without ``Py_GIL_DISABLED``.
+
+
+Motivation
+==========
+
+In its `acceptance post <https://discuss.python.org/t/84319/123>`__
+for :pep:`779`, the Steering Council stated that it “expects that Stable ABI
+for free-threading should be prepared and defined for Python 3.15”.
+
+This PEP proposes how to do this.
+
+
+Background
+----------
+
+Python's Stable ABI, as defined in :pep:`384` and :pep:`652`, provides a way to
+compile extension modules that can be loaded on multiple versions of the
+CPython interpreter.
+Several projects use this to limit the number of *wheels* (binary artefacts)
+that need to be built and distributed for each release, and/or to make it
+possible to test with pre-release versions of Python.
+
+With free-threading builds (:pep:`703`) being on track to eventually become
+the default (:pep:`779`), we need a way to make the Stable ABI available
+to those builds.
+
+To build against the Stable ABI, the extension must use a *Limited API*,
+that is, only a subset of the functions, structures, etc. that CPython
+exposes.
+The limited API is versioned, and building against Limited API 3.X
+yields an extension that is ABI-compatible with CPython 3.X *and above* (modulo
+bugs in CPython).
+Also, the Limited API is not “stable”: its newer versions may remove API that
+was part of its older versions.
+
+This PEP proposes the largest such removal to date.
+
+
+Rationale
+=========
+
+The design in this PEP makes several assumptions:
+
+One ABI
+   A single compiled extension module should support both
+   free-threaded and GIL-enabled builds.
+
+No backwards compatibility
+   The new limited API does not need to support CPython 3.14 and below.
+   Projects that need this can build separate extensions specifically for
+   the 3.14 free-threaded interpreter, and for older stable ABI versions.
+
+API changes are OK
+   The new Limited API may require extension authors to make significant
+   changes to their code.
+   Projects that cannot do this (yet) can continue using Limited API 3.14,
+   which will yield extensions compatible with GIL-enabled builds only.
+
+No extra configuration
+   We do not introduce new “knobs” that influence what API is available
+   and what the ABI is compatible with.
+
+
+Specification
+=============
+
+
+Opaque PyObject
+---------------
+
+Version 3.15 of the Limited API will:
+
+- make the following structures *opaque* (or in C terminology, *incomplete
+  types*):
+
+  - :c:type:`PyObject`
+  - :c:type:`PyVarObject`
+  - :c:type:`!PyModuleDef_Base`
+  - :c:type:`PyModuleDef`
+
+- no longer include the following macros:
```

Comment:
```suggestion
- exclude the following macros:
```

Reply:
I think either wording works, but Petr's rationale makes sense for the current phrasing.